Make the cursorMark based sitemap generation use the search service.

EreMaijala opened the pull request:
This makes it fire events that are used e.g. to handle deduplication. Changes include optimizations to make the process run more or less as fast as before.

I also moved the source identifier injection for records into the collection class, as it's faster than using the iterator interface (around 10% improvement in sitemap generator's throughput).

TODO:
- [x] Commit #1488 first.

EreMaijala committed:
Make the cursorMark based sitemap generation use the search service.

This makes it fire events that are used e.g. to handle deduplication. Changes include optimizations to make the process run more or less as fast as before.

EreMaijala committed:
Try to make PHP 7.1 happy.

demiankatz commented:
@EreMaijala, I think this would benefit from being split into more commits, since there are a lot of distinct purposeful changes that don't necessarily relate directly to one another. At least one of them -- changing the signature of `RecordCollectionInterface::add` -- might have some backward-breaking consequences on custom code and probably needs to wait until release 7... but maybe we can merge most of the other changes right away. If you don't have time, I'm happy to take a stab at merging some of the low-hanging fruit myself, but I'll wait until you're back in town so we can discuss if necessary.

EreMaijala commented:
@demiankatz The changes are only unrelated in the sense that they could be committed separately, but without the signature change that allows the simple record factory the use of search service would cause a significant performance regression. I wouldn't want that to be the case for 6.1, so I think it's better to wait for release 7. Ok, we could get the round() call in separately, but I'm not sure if it's worth it. :)

demiankatz commented:
@EreMaijala, I've put a 7.0 milestone on this so we can deal with it as part of that release. I definitely wouldn't want to introduce a performance regression by merging this piecemeal, but I thought perhaps there were other parts that could be made independent -- for example, does adding getIds support to the DeduplicationListener have a direct dependency on anything else in here? That seemed like it might be an unrelated optimization, but maybe I misunderstand.

EreMaijala commented:
@demiankatz Adding getids support wouldn't cause any trouble, but then it won't be used unless SearchService is used for searching, and I would hesitate to merge code that's not currently used. In the worst case I'd refactor everything and we'd end up with unused code lingering in the code base. :)

demiankatz commented:
Fair enough. But what about the optimization to add the `$checkExisting` flag. Isn't that something we could benefit from right away?

Also, on a separate topic,, if we move away from requiring `RecordInterface` in the collection, we probably need to change a bunch of `@return` comments to reflect the more open interface. I don't entirely like this, since it is useful to be able to identify records in a predictable way. Did you do any experimentation with wrapping the arrays in something extremely lightweight, or is the performance penalty for that still too great?

EreMaijala committed:
Add a SimpleRecord class to keep record collection signature intact.

EreMaijala committed:
Revert optimizations.

EreMaijala commented:
@demiankatz Thanks for your insightful thoughts and persistence! You're absolutely right, and no, I failed to test a very simple record class since I didn't realize how heavy the normal record class really is. A quick test indicates that the performance hit from using a simple record class instead of the array is around 3%, which is not too bad at all.

I moved the optimizations to #1488 and reverted them and the incompatible signature change. I also added a SimpleRecord class just for the sitemap generator.

demiankatz commented:
Thanks, @EreMaijala; I've merged #1488 and will take a closer look at this soon. Reviewing that code makes me wonder if the whole setSourceIdentifier logic needs to be revisited a bit -- as it is, it's a bit confusing, in that if you set the source identifier and then add records subsequently, the new records won't get the right source identifier... a fact which is probably not significant to the code as written, but makes the class interface unintuitive in the abstract. But that's a battle for another day! For now, your improvements certainly don't make that situation any worse. I'm not entirely sure why moving the identifier setting inside the collection class makes a significant performance difference, but it seems cleaner in any case.

EreMaijala commented:
@demiankatz I agree it's confusing. I'd actually rather see the source id injected when the record is created, not when it's added to the collection.

I believe the performance difference comes from avoiding the iterator interface and the associated function calls.

demiankatz commented:
@EreMaijala, I agree that source id injection makes more sense at record creation... though I suspect it was done this way for some specific reason. Might require some digging to figure out the implications of a change. Do you think it's worth opening a JIRA ticket to highlight this as an area for future investigation? I don't think it's an urgent priority right now, but it is something that would be nice to clarify in future.

And yes, that makes sense regarding the iterator interface!

EreMaijala commented:
@demiankatz I've no strong feeling about it. We've managed with the current way for several years now, and I'm ok with leaving it as is and revisit if, it any point, we find it an issue.

demiankatz commented:
Works for me, @EreMaijala!

demiankatz commented on module/VuFind/src/VuFind/Sitemap/SimpleRecord.php:
```diff
@@ -0,0 +1,96 @@
+<?php
+/**
+ * A minimal record class for the sitemap generator
```
There's not really anything Sitemap-specific in here; do you think it might make sense to move this to the VuFindSearch namespace so it's potentially available for other applications as well?

EreMaijala replied:
Sure thing. I just thought that maybe the class is too specialized, but VuFindSearch is probably more logical place anyway.

demiankatz replied:
It doesn't seem too specialized to me -- I'd just change the class-level comment to "A minimal record class for wrapping an array of fields" -- I think that might be useful in a variety of situations.

EreMaijala replied:
Ok, I moved SimpleRecord.

demiankatz replied:
Thanks!

demiankatz commented on module/VuFind/src/VuFind/Sitemap/Generator.php:
```diff
@@ -347,13 +370,18 @@ protected function getIdsFromBackendUsingCursorMark(Backend $backend, $cursorMar
+            0,
+            $this->countPerPage,
+            $params,
+            'getids'
```
If our goal is to standardize this so that standard listeners fire on sitemap generation, using this non-standard context seems likely to cause issues, since most listeners are going to be expecting `search` in this context. I understand that you need a way to flag this behavior for the DeduplicationListener; I just wonder if we need a better mechanism to accomplish that. For example, would stuffing something into the ParamBag potentially be less disruptive? Or do we need some new way of tagging requests independently of the ParamBag?

EreMaijala replied:
As I see it this is a new context (fetching ID's instead of doing a normal search), and using 'getids' is more back-compatible since the listeners weren't firing before either. Any existing listener wouldn't know to check a new flag in parameters. Yes, the downside is that existing listeners need to be modified for getids, but that makes sure they're also modified only if appropriate.

demiankatz replied:
That makes sense, but if we go that route, I wonder if we should go a little further and actually implement a getids method on the backend. Running a search under an alias like this seems like a bit of a hack, and I think it makes for somewhat non-obvious code. We could implement this as an optional Feature interface, like RetrieveBatchInterface, etc., so we don't have to implement it across all the backends.

It's entirely possible this is more work than it's worth -- I imagine this could end up duplicating a lot of code and/or creating IdCollections and IdCollectionFactories parallel to RecordCollections, etc. -- but I just want to talk it through and see if we can do this in a more robust and straightforward way before settling on this approach.

EreMaijala replied:
I added the feature interface and a getIds method that uses a normal search as fallback. I didn't tie the different collection type to it, since it can just as well be used with the normal record collection if need be. What do you think?

demiankatz replied:
Sounds reasonable to me; I'll give this a closer review next week when I have a bit more time!

EreMaijala committed:
Refactor getIds to a feature interface. Move SimpleRecord.

demiankatz commented on module/VuFind/src/VuFind/Sitemap/Generator.php:
```diff
@@ -331,12 +344,21 @@ protected function getIdsFromBackendUsingCursorMark(Backend $backend, $cursorMar
 
+        // Set up the record factory on first call. We use a very simple factory
+        // since performance is important and we only need the identifier.
+        if ('*' === $cursorMark) {
```
Would it be cleaner to have a `configureBackend` method that gets called at the top of `generateForBackend` and does this work conditionally based on `$this->retrievalMode` rather than basing it on the cursor mark, which seems less clear and more potentially error prone?

EreMaijala replied:
Yes. I made it setupBackend.

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Service.php:
```diff
@@ -375,6 +412,38 @@ function ($o) {
+     *
+     * @return RecordCollectionInterface
+     */
+    protected function performSearch($backend, Query\AbstractQuery $query,
```
Is this performSearch method actually used for anything? I don't see any calls to it, but maybe I'm overlooking something.

EreMaijala replied:
removed

demiankatz commented on module/VuFindConsole/src/VuFindConsole/Controller/UtilController.php:
```diff
@@ -282,7 +282,9 @@ public function sitemapAction()
         $generator = new Sitemap(
             $this->serviceLocator->get(\VuFind\Search\BackendManager::class),
-            $configLoader->get('config')->Site->url, $configLoader->get('sitemap')
+            $this->serviceLocator->get(\VuFindSearch\Service::class),
```
I wonder if it's worth setting up a factory for the Sitemap class rather than doing all this work inline in the controller. Certainly not a requirement right now, but maybe a change worth considering while we're touching things anyway.

EreMaijala replied:
...and this was done a couple of weeks ago.

demiankatz reviewed (commented):
Thanks again, @EreMaijala -- I had time for a closer review, and a few more questions came up. Once you've had a chance to address them, I'll take a little more time for hands-on testing.

EreMaijala committed:
Add separate backend setup method and generator factory.

EreMaijala committed:
Merge commit '316c9b8fae6fb0bee39ddc159e208aedb817a8e8' into sitemap-events

EreMaijala commented:
Resolved conflicts. This should be ready to go when the time is right.

demiankatz commented on module/VuFindSearch/src/VuFindSearch/Response/SimpleRecord.php:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * Source backend identifier
+     */
+    protected $sourceId = 'Solr';
```
Any objection to using DEFAULT_SEARCH_BACKEND here in place of the hard-coded Solr?

EreMaijala replied:
No objection, fixed.

demiankatz reviewed (commented):
Thanks for the progress, @EreMaijala! This is looking good, apart from one last minor question. If you don't mind that change, I'll do some double-checking that everything is working correctly, and then we should be able to merge this (though I may wait until 7.0, since these are fairly significant structural changes for a minor release, and we're nearly done with 6.1 anyway).

EreMaijala committed:
Use DEFAULT_SEARCH_BACKEND instead of harcoded 'Solr'.

demiankatz committed:
Merge remote-tracking branch 'origin/master' into sitemap-events

demiankatz commented on module/VuFindConsole/config/module.config.php:
```diff
@@ -42,6 +42,7 @@
     ],
     'service_manager' => [
         'factories' => [
+            'VuFind\Sitemap\Generator' => 'VuFind\Sitemap\GeneratorFactory',
```
@EreMaijala, this branch is not working for me, because this GeneratorFactory is missing. Did you neglect to commit the file to this branch? Thanks!

demiankatz committed:
Add test coverage for getIds.

demiankatz committed:
Add test coverage for SimpleRecord.

demiankatz committed:
More test coverage.

demiankatz committed:
More test coverage; corrected return value.

demiankatz commented:
@EreMaijala, I did some work to add test coverage for the new VuFindSearch features in order to keep our coverage up. Once you commit the missing factory, I'll do some final testing and hopefully get this merged to master.

demiankatz committed:
php-cs-fixer

EreMaijala committed:
Add missing factory.

EreMaijala commented:
@demiankatz Oops, factory added. Thanks for adding the tests!

demiankatz committed:
Merge branch 'master' into sitemap-events

demiankatz committed:
Merge branch 'master' into sitemap-events

demiankatz commented:
Thanks, @EreMaijala -- I've manually merged some of the underlying infrastructure pieces to master separately to keep the commit history cleaner, and have now merged this PR as well. Thanks!
